### PR TITLE
Verify the index state of concrete indices after alias resolution

### PR DESCRIPTION
**Pull request by martijnvg**

### CODE REVIEW

File: src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 1]
```diff
@@ -760,24 +760,49 @@ public String concreteSingleIndex(String indexOrAlias, IndicesOptions indicesOpt
...
+        }
+
+        switch (indices.length) {
+            case 0:
+                return indices;
+            case 1:
+                indexMetaData = this.indices.get(indices[0]);
+                if (indexMetaData != null && indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+                    if (failClosed) {
+                        throw new IndexClosedException(new Index(aliasOrIndex));
```
Comment (javanna) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: hey @martijnvg I double checked with @clintongormley and we both think it's better to add the actual index that was closed, not the alias. Knowing that an alias is closed has little sense, better to report back which concrete index was closed.

File: src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 2]
```diff
@@ -760,24 +760,49 @@ public String concreteSingleIndex(String indexOrAlias, IndicesOptions indicesOpt
         }
         // not an actual index, fetch from an alias
         String[] indices = aliasAndIndexToIndexMap.getOrDefault(aliasOrIndex, Strings.EMPTY_ARRAY);
-        if (indices.length == 0 && !allowNoIndices) {
+        if (indices.length == 0 && failNoIndices) {
```
Comment (javanna) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Can you double check the javadocs and es docs around indices options? Those options affect now not only how wildcards get expanded, but also how aliases get resolved to concrete indices.

File: src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 3]
```diff
@@ -760,24 +760,51 @@ public String concreteSingleIndex(String indexOrAlias, IndicesOptions indicesOpt
...
+                    if (indexMetaData != null) {
+                        if (indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+                            if (failClosed) {
+                                throw new IndexClosedException(new Index(aliasOrIndex));
+                            } else if (!options.forbidClosedIndices()) {
+                                concreteIndices.add(index);
+                            }
+                        } else if (indexMetaData.getState() == IndexMetaData.State.OPEN) {
+                            concreteIndices.add(index);
+                        }
```
Comment (javanna) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: nitpicking: maybe add an else that throws exception (index state not supported?)

**Comment by javanna** [comment]
Hey @martijnvg I left a couple of comments around docs and the returned error message, if you can address those then this is ready to be merged ;)

File: src/main/java/org/elasticsearch/cluster/metadata/MetaData.java (hunk 1 above)
Comment (martijnvg) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Ok, will change!

**Comment by martijnvg** [comment]
@javanna I've updated the PR.

File: src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 4]
```diff
@@ -760,24 +760,53 @@ public String concreteSingleIndex(String indexOrAlias, IndicesOptions indicesOpt
...
-            }
+                return indices;
+            default:
+                ObjectArrayList<String> concreteIndices = new ObjectArrayList<>();
+                for (String index : indices) {
+                    indexMetaData = this.indices.get(index);
+                    if (indexMetaData != null) {
+                        if (indexMetaData.getState() == IndexMetaData.State.CLOSE) {
+                            if (failClosed) {
+                                throw new IndexClosedException(new Index(aliasOrIndex));
```
Comment (javanna) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: name needs to be replaced here...that is why I don't like these optimizations :) they make the code error-prone ;)

Comment (martijnvg) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: argg... i'll change it :)

**Comment by javanna** [comment]
LGTM

**Pull request merged**